module/os/linux/zfs/zfs_znode.c

```diff
@@ -606,17 +606,24 @@ zfs_znode_alloc(zfsvfs_t *zfsvfs, dmu_buf_t *db, int blksz,
 	 * number is already hashed for this super block.  This can never
 	 * happen because the inode numbers map 1:1 with the object numbers.
 	 *
-	 * The one exception is rolling back a mounted file system, but in
-	 * this case all the active inode are unhashed during the rollback.
+	 * Exceptions include rolling back a mounted file system, either
+	 * from the zfs rollback or zfs recv command.
+	 *
+	 * Active inodes are unhashed during the rollback, but since zrele
+	 * can happen asynchronously, we can't guarantee they've been
+	 * unhashed.  This can cause hash collisions in unlinked drain
+	 * processing so do not hash unlinked znodes.
 	 */
-	VERIFY3S(insert_inode_locked(ip), ==, 0);
+	if (links > 0)
+		VERIFY3S(insert_inode_locked(ip), ==, 0);
 
 	mutex_enter(&zfsvfs->z_znodes_lock);
 	list_insert_tail(&zfsvfs->z_all_znodes, zp);
 	zfsvfs->z_nr_znodes++;
 	mutex_exit(&zfsvfs->z_znodes_lock);
 
-	unlock_new_inode(ip);
+	if (links > 0)
+		unlock_new_inode(ip);
```

Comment:
Actually, one more question.  What's the rational behind not unlocking the inode here?  Since it's unhashed it should be undiscoverable (which is what we want), but I believe we still need to call `unlock_new_inode()` to clear `I_NEW` otherwise in `iput_final()` we'll [generate a warning](https://elixir.bootlin.com/linux/v5.13-rc1/source/fs/inode.c#L1627) about `I_NEW` still being set.

Reply:
Did a little investigation.  zfs_znode_alloc -> new_inode leaves i_state at 0, not I_NEW.  So we are ok here with that.

```
        if (links > 0)
                unlock_new_inode(ip);
+       else
+               zfs_dbgmsg("inode->i_state & I_NEW: %d i_state %d",
+                   (ip->i_state & I_NEW), ip->i_state);
>>>
Fri Jun 11 17:12:39 2021   zfs_znode.c:628:zfs_znode_alloc(): inode->i_state & I_NEW: 0 i_state 0
```

Reply:
That addresses my last concern.  Thanks for digging in to this.